File: backend/app/models/projections.py
from typing import Optional

from pydantic import BaseModel

from app.models.response import ApiResponse


class ActualMonth(BaseModel):
  """Actual traded results for a single calendar month of the current year."""

  month: int
  label: str
  cumulative_r: float
  month_r: float
  trades_count: int


class ProjectedMonth(BaseModel):
  """Monte Carlo percentile fan for a single future calendar month."""

  month: int
  label: str
  p10: float
  p25: float
  p50: float
  p75: float
  p90: float
  estimated_trades: int


class ProjectionStats(BaseModel):
  """Descriptive statistics over the historical trade pool."""

  expectancy: float
  win_rate: float
  std_deviation: float
  skewness: float
  kurtosis: float
  total_trades: int
  best_trade: float
  worst_trade: float
  max_winning_streak: int
  max_losing_streak: int


class GoalResult(BaseModel):
  """Probability of reaching a cumulative-R target by year-end."""

  target_r: float
  probability: float


class RiskResult(BaseModel):
  """Downside risk metrics derived from the Monte Carlo paths."""

  ruin_probability: float
  max_drawdown_median: float


class FiltersApplied(BaseModel):
  """The query filters that were active for this projection run."""

  start_date: Optional[str] = None
  assets: list[str]


class Projections(BaseModel):
  """Full projection payload returned by GET /api/projections."""

  actual_months: list[ActualMonth]
  projected_months: list[ProjectedMonth]
  stats: ProjectionStats
  goal: Optional[GoalResult] = None
  risk: RiskResult
  filters_applied: FiltersApplied


class ProjectionsResponse(ApiResponse[Projections]):
  """Standard { data, error } envelope around the projections payload."""

File: frontend/src/types/projections.ts
/**
 * Domain types for the projections endpoint.
 *
 * Field names mirror the API's snake_case JSON — same convention as
 * `@/types/analytics`. The response wrapper type represents the unwrapped
 * `data` payload (the api client already strips the `{data,error}` envelope).
 */

// ---------------------------------------------------------------------------
// Filter params
// ---------------------------------------------------------------------------

export interface ProjectionFilters {
  start_date?: string
  assets?: string[]
  goal_r?: number
}

// ---------------------------------------------------------------------------
// API shapes — mirror backend/app/models/projections.py exactly
// ---------------------------------------------------------------------------

export interface ActualMonth {
  month: number
  label: string
  cumulative_r: number
  month_r: number
  trades_count: number
}

export interface ProjectedMonth {
  month: number
  label: string
  p10: number
  p25: number
  p50: number
  p75: number
  p90: number
  estimated_trades: number
}

export interface ProjectionStats {
  expectancy: number
  win_rate: number
  std_deviation: number
  skewness: number
  kurtosis: number
  total_trades: number
  best_trade: number
  worst_trade: number
  max_winning_streak: number
  max_losing_streak: number
}

export interface GoalResult {
  target_r: number
  probability: number
}

export interface RiskResult {
  ruin_probability: number
  max_drawdown_median: number
}

export interface FiltersApplied {
  start_date: string | null
  assets: string[]
}

export interface Projections {
  actual_months: ActualMonth[]
  projected_months: ProjectedMonth[]
  stats: ProjectionStats
  goal: GoalResult | null
  risk: RiskResult
  filters_applied: FiltersApplied
}
